src/ucp/core/ucp_worker.c

```diff
@@ -1992,7 +2011,7 @@ static void ucp_worker_keepalive_reset(ucp_worker_h worker)
     worker->keepalive.timerfd     = -1;
     worker->keepalive.cb_id       = UCS_CALLBACKQ_ID_NULL;
     worker->keepalive.last_round  = 0;
-    worker->keepalive.lane_map    = 0;
+    worker->keepalive.check_ep    = NULL;
```

Comment:
1. do we really need check_ep now? can't we use just "iter" to point to current ep?
2. looks like we don't really use the value of "check_ep", but only compare it to NULL. so if really needed, it could be an int or a flag

Reply:
what additional state is captured by check_ep, that is not covered by ep iterator?

Reply:
if we should go to the next EP, we can't somehow indicate it through iterator

Reply:
maybe the iterator can always point to the ep we should try to send on?
if we didn't send, don't advance the iterator. and skip eps that should not send.

Reply:
> maybe the iterator can always point to the ep we should try to send on?

it is already true. we stop doing keepalive if reach `&worker->all_eps`.

> if we didn't send, don't advance the iterator. and skip eps that should not send.

we already skip EPs which have `FAILED` flag

Reply:
so why need to indicate "we should go to next ep"? what does it mean? I would expect we always try current ep, and then go to next ep..

Reply:
> so why need to indicate "we should go to next ep"?

if `uct_ep_check`/`uct_ep_am_bcopy` returns `UCS_ERR_NO_RESOURCE` we shouldn't "go to the next ep"

Reply:
right, in this case the iterator should not advance